kategory-effects/src/main/kotlin/kategory/effects/typeclass/MonadErrorCancellableContinuations.kt

```diff
@@ -0,0 +1,57 @@
+package kategory.effects
+
+import kategory.*
+import kategory.effects.data.internal.BindingCancellationException
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlin.coroutines.experimental.CoroutineContext
+import kotlin.coroutines.experimental.EmptyCoroutineContext
+import kotlin.coroutines.experimental.RestrictsSuspension
+import kotlin.coroutines.experimental.intrinsics.COROUTINE_SUSPENDED
+import kotlin.coroutines.experimental.intrinsics.suspendCoroutineOrReturn
+import kotlin.coroutines.experimental.startCoroutine
+
+typealias Disposable = () -> Unit
+
+@RestrictsSuspension
+open class MonadErrorCancellableContinuation<F, A>(ME: MonadError<F, Throwable>, override val context: CoroutineContext = EmptyCoroutineContext) :
+        MonadErrorContinuation<F, A>(ME) {
+
+    protected val cancelled: AtomicBoolean = AtomicBoolean(false)
+
+    fun disposable(): Disposable = { cancelled.set(true) }
+
+    internal fun returnedMonad(): HK<F, A> = returnedMonad
+
+    override fun resumeWithException(exception: Throwable) {
+        returnedMonad = ME.raiseError(exception)
+    }
+
+    override suspend fun <B> bind(m: () -> HK<F, B>): B = suspendCoroutineOrReturn { c ->
+        val labelHere = c.stackLabels // save the whole coroutine stack labels
+        returnedMonad = flatMap(m(), { x: B ->
+            c.stackLabels = labelHere
+            if (cancelled.get()) {
+                throw BindingCancellationException()
+            }
+            c.resume(x)
+            returnedMonad
+        })
+        COROUTINE_SUSPENDED
+    }
+}
+
+/**
+ * Entry point for monad bindings which enables for comprehensions. The underlying impl is based on coroutines.
+ * A coroutines is initiated and inside `MonadErrorContinuation` suspended yielding to `flatMap` once all the flatMap binds are completed
+ * the underlying monad is returned from the act of executing the coroutine
+ *
+ * This one operates over MonadError instances that can support `Throwable` in their error type automatically lifting
+ * errors as failed computations in their monadic context and not letting exceptions thrown as the regular monad binding does.
+ *
+ * This operation is cancellable by calling invoke on the [Disposable] return.
```

Comment:
The type of excection raised to the context of the Monad should be documented as the method signature just states `Throwable`.

Reply:
Added!